WIP: enhancement proposal: "Add Providers Support" (#2272)
openshift/assisted-service: docs/enhancements/add-external-providers.md

---
title: Add-Providers-Support
authors:
  - "@Gal-Zaidman"
creation-date: 2021-07-25
last-updated: 2021-07-25
---

# Add Providers Support

## Summary

> Currently, the assisted installer supports Bare-Metal installations only.

Reviewer: vSphere is already supported (in 'master' branch at least)

Author: well, not this way at least...
I mean the code is in I know that, but we defined criteria for testing, provider detection, documentation, and so on...
I don't have a problem with adding it as supported by the assisted installer if you feel strongly about it...

Reviewer 2: The solution here may be to find more precise language than "supports". AI enables you to install on any platform where you can boot the discovery ISO. But I think what you're trying to get at here is that the resulting cluster's formal "platform" is limited to baremetal or none.

Author: - I have no strong feelings against including VSphere in this description so I will add it.
- @mhrivnak added your suggestions as well

Openshift can be installed on various providers such as:
- Red Hat Virtualization.
- OpenStack.

> - VSphere.

Reviewer: ```suggestion
- vSphere.
```

- AWS.
- GCP.

> - Azure.

Reviewer: We don't intend on supporting cloud installations, just on-prem.

Author: Well yes but I'm just stating here that Openshift can be installed on them.
If you feel it should be mentioned I can add it into the Non-Goals, but since this enhancement is for adding the provider interface, add oVirt and migrate existing providers BM and vSphere I don't think it is required.

Reviewer 3: Not sure what do you mean by migrate....

Reviewer 4: Perhaps we can use some wording like:

"OpenShift has an interface to select the target provider, which is currently not being used by assisted-service (visit this link to learn more about OCP's provider support).

This enhancement will focus on exposing such interface through assisted-installer in order to make it easier to enable more providers in the future"

Probably needs some tweaks but you get the idea :)

Author: - @romfreiman I meant move the current parts of the code (for example vSphere support that was recently added) to the provider interface.

- @avishayt again since it's just the summery I have no strong feeling here, but I don't think it hurts stating that Opeshift supports also cloud providers... we are talking about a provider interface. I will add it as a non-goal to make it clearer.
- @flaper87 I think that will make it a lot more complex than it should be since provider support is much bigger than the installer, and the installer project doesn't really have such an interface :).

Reviewer 5: The question "which platforms does assisted work on" has caused a lot of confusion across openshift engineering in the past. The specific question "but why would I want to use assisted on AWS" has definitely come up a lot, the answer was, or at least from what I understand, is that you wouldn't want to because we have IPI there and it works well for that scenario.

To reduce potential for confusion in this doc, I would recommend not listing the platforms we know RH doesn't intend to use assisted on, else you open yourself to those kinds of questions again

Author: OK edited

This enhancement will define a clear interface to extend the assisted installer and allow
provider installations.

## Motivation

Assisted Installer can be a great tool to perform UPI (User Provisioned) installation on external providers,
providing the user with clear UI to see its provisioned machines joining the cluster, and lowering installation
resources requirements.

Assisted Installer is also ideal for mixed env configurations.
** The RHV provider supports installing a mixed env where the Control-Plain (Master) nodes are deployed on top of RHV
and Compute (Worker) nodes are Bare-Metal nodes.

### Goals

- Create a Provider interface, and hooks in the assisted installer code which separates the assisted installer core code
 from the provider-specific code.
- Add clear documentation for each provider specific function.
- Document steps to install a provider cluster with assisted installer

### Non-Goals

- Support IPI installations.
- Integrating all relevant [Openshift installer types](https://github.com/openshift/installer/tree/master/pkg/types)
into the assisted installer code - this will require a large refactor of the assisted installer code.

> - Remove Baremetal from not appropriate stuct/functions names(example in open questions).

Reviewer: ```suggestion
- Remove Baremetal from not appropriate struct/function names (example in open questions).
```

 
## Proposal

### User Stories

#### Story 1

> As a RHV/VSphere admin I want to create an Openshift cluster on my RHV engine by using the assisted installer.

Reviewer: ```suggestion
As a RHV admin I want to create an Openshift cluster on my RHV engine by using the assisted installer.
```

#### Story 2

As a RHV admin I want to create an Openshift cluster on my RHV engine by using the assisted installer in which

> Control-Plain (Master) nodes are deployed on top of RHV and Compute (Worker) nodes are Bare-Metal nodes.

Reviewer: @eliorerz Note this will change the logic of the get-supported-hosts API.

Reviewer 3: @avishayt hopefully not. The get-supported-hosts api is should provide another interface per platfrom, right? So every platform will have different logic

Reviewer 6: Is this (multiple platforms in a single cluster) something the openshift installer support?

Author: Also see comment: https://github.com/openshift/assisted-service/pull/2272/files#r677255809

#### Story 3

As an Openshift developer working on platform X I want to add support for installing Openshift with assisted installer
on top of platform X.

### Implementation Details/Notes/Constraints

#### Changes to the assisted-service project:

##### Step 1: Detect platform specific logic

> Go over the assisted installer flow and detect parts of the code which require provider specific logic.

Reviewer: Today we rely on the "Manufacturer" that is returned by our inventory

Author: I didn't mean to detect the platform (this is step 4).
Here I meant to identify parts of the code that should be under the provider interface

Reviewer: I see, a research step.  OK.

##### Step 2: Create a Go interface

We should create a Go interface that will contain each provider specific logic in a function
for each provider to implement.

For example, prototype:

```go
// Provider contains functions which are required to support installing on a specific platform.
type Provider interface {
	// Name returns the name of the platform.
	Name() string
	// UpdatePlatformInventoryData sets all platform specific inventory data on the updates map.
	UpdatePlatformInventoryData(params installer.UpdateClusterParams, cluster *common.Cluster, updates map[string]interface{}) error
    // SetPlatformInstallConfig sets all platform specific install config parameters

>     SetPlatformInstallcfgValues(platform *platformVsphere, clusterPlatform *models.VspherePlatform)

Reviewer: Fix indentation

    //...
}
```

##### Step 3: Implement registry

Implement the registry pattern so providers can register with their names.
When we will detect the a certain provider then we will register it.

> If no provider is detected then we register the Bare-Metal provider.

Reviewer: This section is unclear.  Is this describing how the service understands which platforms are available, or how the user chooses which platform to use?  I assume the first.
We will have the bare-metal and none platforms in any case (even if the user is running on virt, in case they don't want/need platform integration).

Author: Neither of them actually.
The "registry pattern" is a design pattern that is used a lot in openshift.
Here is an example from the api-server:
- Plugins Register function: https://github.com/openshift/openshift-apiserver/blob/98786f917ffc7d3dc3b05893f405970b87a419b9/vendor/k8s.io/apiserver/pkg/admission/plugins.go#L70-L86
- Usage: https://github.com/openshift/openshift-apiserver/blob/98786f917ffc7d3dc3b05893f405970b87a419b9/pkg/build/apiserver/admission/secretinjector/admission.go#L24-L30

Here is an example that is closer to our usecase, from a PR we abandoned:
- Register definition: https://github.com/openshift/installer/pull/4517/commits/11dc0bcf5555caea832cda52f6c52b64590f6de6#diff-dbd907695a0fe9e8040355b09537c81ac2afd21ad450a2e16468b495ac9d862dR34-R36
- Usage: https://github.com/openshift/installer/pull/4517/commits/21c2ec8c0a1f1450da404083981b5c47d829a3a2#diff-cb477a5c4cf6ed9e8ca2b7c1e817d98737d8bda0879cd30182129b2cc44c6ffeR10

it will be used inside the providerDetector method to register the provider once it is detected

Reviewer: I think that's what I meant by "describing how the service understands which platforms are available"?
Each provider registers its name and some method that the service can execute to determine if the platform is available for the current hosts that are discovered.  For example, the RHV plugin with register "RHV" and a method that checks that all of the masters are running on RHV.  At that point, if the method returns True, the service will allow the user to select that platform.  If the user selects it, then the other plugin methods can be executed, such as validations, configuration, etc.

Author: Amm yep that is what I meant, one thing that is not clear to me (due to my lack of AI project knowledge probably) how will the user select the oVirt platform if there is no UI changes?
BTW I think we can move this to Open Question 2 since it is related to it

##### Step 4: Implement providerDetector method

> Implement a function which determines which provider should be used.

Reviewer: ```suggestion
Implement a function which determines which providers can be used.
```

Author: Regarding the can/should I think that can be the provider's decision on how to implement, but it is an interesting question to ask.
If there is a use case in which the user will provision the Nodes on a certain provider but want to disable provider integration? I'll add that to the open questions

Reviewer: Sure.  We see that today.  Maybe they don't have platform credentials, or they are doing some custom integrations.

Each provider should implement its own detection logic.
For example:
if all control plain nodes are running on oVirt
(i.e have [SystemVendor.Manufacturer](https://github.com/openshift/assisted-service/blob/master/models/system_vendor.go)

> == oVirt OR RHV) then register the oVirt provider.

Reviewer: ```suggestion
== oVirt OR RHV) then recommend the oVirt provider.
```

Author: See comment above, added to open questions

Author: Changed

##### Step 5: Add calls to the provider functions

> When ever we need to call some provider specific function we should use the provider interface instead of a switch case.

Reviewer: ```suggestion
Whenever we need to call some provider-specific function we should use the provider interface instead of a switch case.
```

We need to make sure that the provider is listed in the 
[isVirtual](https://github.com/openshift/assisted-installer-agent/blob/master/src/inventory/system_vendor.go#L15-L19]
list and that it can be detected using the fields in 
[SystemVendor](https://github.com/openshift/assisted-service/blob/master/models/system_vendor.go).

#### Changes to the Openshift installer project:

A small change is required in the installer project.
Since the openshift installer is generating a random InfraID the names of the Nodes(in the manifest) are different
from the nodes the user provisions, and since we have a provider set then the provider will try to start new machines 
due to the naming mismatch.
Since the InfraID is random we can't tell the user to create the Nodes with specific names.
Also, the InfraID is later used by the platforms in various ways to identify the cluster,
for example in oVirt we use it to set an oVirt tag which helps us group the cluster resources in oVirt,
and we rely on it in various places such as cluster destroy or CSI/Machine provider logic to filter cluster resources
quickly and prevent unnecessary API calls.

> We need to add an env var to allow us to override the generated InfraID.

Reviewer 7: The generation of the infraID is random but it can be known as in AWS UPI: https://github.com/openshift/installer/blob/master/docs/user/aws/install_upi.md#extract-infrastructure-name-from-ignition-metadata

Extracting the infra id would be preferred if it satisfies the needs. If it does not, let's discuss why.

Author: In the assisted installer the machines are provisioned before the ignition-configs or manifests are generated.
If we will go with this approach we will need to make manual modifications to the User Provisioned Nodes

Reviewer: How does it work with UPI?

Reviewer 6: Or we can chagne to the ignition config rendered by the installer :wink:

Author: @eranco74 it also affects the manifests a lot.
@avishayt On current RHV UPI we don't enable provider integration, we have a separate effort to document this as well, but essentially since the manifests and ignition can be created before there are any Nodes there is no problem with extracting the infraID and then creating the VMs with that infraID in the name

---
title: Add-Providers-Support
authors:
  - "@Gal-Zaidman"
creation-date: 2021-07-25
last-updated: 2021-07-25
---

# Add Providers Support

## Summary

Currently, the assisted installer supports Bare-Metal installations only.
Openshift can be installed on various providers such as:
- Red Hat Virtualization.
- OpenStack.
- vSphere.
- AWS.
- GCP.
- Azure.

> This enhancement will define a clear interface to extend the assisted installer and allow

Reviewer 2: Somewhere in this proposal, it would be valuable to elaborate on what kinds of provider-specific behaviors will be addressed. Including a list of all such behaviors we're aware of so far would be useful. That'll give people a more tangible sense of what to expect.

Author: I aimed for the [provider interface](https://github.com/openshift/assisted-service/pull/2272/files#diff-988c4e5d655472b2174837171f5d89978ca0f86ec8554329b386dccb9990ea83R84-R90) to do that.
I would expand it and update the enhancement as we will find more provider-specific parts, in the meantime, I relaied on the adding vSphere PR that was merged by the team

provider installations.

## Motivation

> Assisted Installer can be a great tool to perform UPI (User Provisioned) installation on external providers,

Reviewer 4: ```suggestion
Assisted Installer can be a great tool to perform User Provisioned installation (UPI) on external providers,
```

providing the user with clear UI to see its provisioned machines joining the cluster, and lowering installation
resources requirements.

Assisted Installer is also ideal for mixed env configurations.

> ** The RHV provider supports installing a mixed env where the Control-Plain (Master) nodes are deployed on top of RHV

Reviewer 2: s/Plain/Plane

> and Compute (Worker) nodes are Bare-Metal nodes.

Reviewer 8: How do we plan to handle the expansion of the test matrix here?  Would we support every provider for single-provider and mixed-provider, or only $provider and baremetal (as is possible today due to the hard-coded `baremetal` platform in AI)

A common requirement is to deploy controlplane nodes on some kind of VM platform, then deploy worker hosts on BM (potentially in an automated way using Metal3), but if we enable it via this AI API, how will we ensure all-the-combinations actually work?

So far we've deferred adding a generic [CBO](https://github.com/openshift/cluster-baremetal-operator) API due to this test-matrix concern - in particular network configuration of controlplane hosts to enable PXE may be different depending on the controlplane provider, etc?

I also wonder about the customer experience vs IPI - e.g in this case you deploy on RHV and have to make a day-1 choice to only ever deploy either RHV or baremetal workers.  So probably customers will then ask for a combination of both, which isn't possible via the current MAPI integration in OCP /cc @JoelSpeed

Author: > How do we plan to handle the expansion of the test matrix here? Would we support every provider for single-provider and mixed-provider, or only $provider and baremetal (as is possible today due to the hard-coded `baremetal` platform in AI)

Amm not sure what test matrix you are referring to, to my knowledge RHV is the only provider that allows such mix env, meaning have platform integration (the ovirt provider and csi driver) but still have some BM workers and some oVirt workers.
So I don't think it would affect other providers.

> A common requirement is to deploy controlplane nodes on some kind of VM platform, then deploy worker hosts on BM (potentially in an automated way using Metal3), but if we enable it via this AI API, how will we ensure all-the-combinations actually work?

I think what you mentioned above is archived with setting platform to none, which means no provider integration and in that case we support "Installing a cluster on any platform" flow.
In RHV we allow a unique case in which we have full-provider integration but allow for non RHV nodes.

> So far we've deferred adding a generic [CBO](https://github.com/openshift/cluster-baremetal-operator) API to enable Metal3 on any platform due to this test-matrix concern - in particular network configuration of controlplane hosts to enable PXE may be different depending on the controlplane provider, etc?
> 
> I also wonder about the customer experience vs IPI - e.g in this case you deploy on RHV and have to make a day-1 choice to only ever deploy either RHV or baremetal workers. So probably customers will then ask for a combination of both, which isn't possible via the current MAPI integration in OCP /cc @JoelSpeed

On RHV this is possible, due to the fact that RHV provider is a part of the machine-controller.
This is the [PR which allowed it](https://github.com/openshift/cluster-api-provider-ovirt/pull/95) and this is the [BZ](https://bugzilla.redhat.com/show_bug.cgi?id=1929702)

Author: It is something that apparently has become pretty common on RHV and a good use-case for the assisted installer for us.
You can do it with IPI or UPI it doesn't matter, and using the AI with platform set to ovirt will allow users to see their bare metals joining the OCP on RHV cluster

Reviewer 8: > Amm not sure what test matrix you are referring to

I mean which combinations will we actually test and expect to support?  Adding flexibility to mix platforms means we'll have to test the combinations we expect to support, and find some way (at least via docs) to communicate the supported combinations to users.

Reviewer: I think specific platforms or combinations thereof are out of scope here.  This is talking about creating some framework for easily adding platforms.  Before any platform is added, there must be approval from PM, QE, etc. so that it will be supported.

Reviewer 9: @avishayt - Let's not confuse this open source contribution with what we'll end up supporting in our downstream product or service offerings. The former does not require any approval apart from the maintainers of this project - if it doesn't break other flows, if it's aligned with the project goals, if it's technically sound, it should go in.

Reviewer: Although Assisted Installer is open source, there's no community around it yet and nobody else is using it.  There's no reason to include code which we won't test and won't use.

Author: @hardys the only added combinations here are OCP on oVirt combinations:
- with mixed env (since only oVIrt is supporting it).
- All nodes are oVirt vms.

@avishayt you are right that to get it downstream we will need PM and QE approval, and we are working on that front.
Not sure how assisted installer U/S and D/S work is being handled and if it is possible to merge something into the assisted service without it going into D/S... but this complication may not be required.
Since a lot of the vSphere work is in, we can get the interface and port the vSphere work to use it and after that add oVirt provider (I assume that by then we will have the approval we need) and this should be an easier part.

Reviewer 2: > Let's not confuse this open source contribution with what we'll end up supporting in our downstream product or service offerings. The former does not require any approval apart from the maintainers of this project - if it doesn't break other flows, if it's aligned with the project goals, if it's technically sound, it should go in.

@Gal-Zaidman as is typical with OSS projects, a key criteria for merging something big and new, like support for a new platform, is having people in-place who are committed to maintaining it. If we expect RH engineers to maintain support for a platform in this project upstream, then we need RH alignment on committing that engineering time. The same would apply to any other contributor proposing to add a big new feature; they would need to demonstrate that there are people able and willing to maintain it in the upstream project.

You are correct that the maintainers upstream can merge what they see fit, and that doesn't have to directly impact downstream. But responsible upstream maintainers will ensure that a contribution has sufficient ongoing support.

Reviewer 5: > On RHV this is possible, due to the fact that RHV provider is a part of the machine-controller.
This is the PR which allowed it and this is the BZ

I don't think this is what Steve was referring to in his comment.

 MAO deploys machine controllers based on the value of the `Infrastructure` resource `status.platformStatus.type`. When installing, the customer will set this to the main topology for the cluster, I expect to match the control plane. In the case of an RHV control plane, we expect the oVirt Machine Controller to be deployed, but not the baremetal machine controller.

So any Machine that is created, is expected to be an oVirt machine. A customer might expect to be able to create an oVirt machine and a baremetal machine, but the oVirt controller is the only one running, and it won't understand the baremetal machines so will error on them constantly.

Machine API isn't designed to work multi platform today, so I don't think there's a way we can deploy both oVirt and baremetal machine controllers simultaneously and expect them to work

Author: I think you meant to tag @mykaul
> @Gal-Zaidman as is typical with OSS projects, a key criteria for merging something big and new, like support for a new platform, is having people in-place who are committed to maintaining it. If we expect RH engineers to maintain support for a platform in this project upstream, then we need RH alignment on committing that engineering time. The same would apply to any other contributor proposing to add a big new feature; they would need to demonstrate that there are people able and willing to maintain it in the upstream project.
> 

Yes I agree, but this is something that you already have since the OCP on RHV team agreed to maintain the code and CI for the effort

> You are correct that the maintainers upstream can merge what they see fit, and that doesn't have to directly impact downstream. But responsible upstream maintainers will ensure that a contribution has sufficient ongoing support.

Reviewer 9: We'd be more than happy to maintain it. This is applicable for both RHV and OSP.

Author: > > On RHV this is possible, due to the fact that RHV provider is a part of the machine-controller.
> > This is the PR which allowed it and this is the BZ
> 
> I don't think this is what Steve was referring to in his comment.
> 
> MAO deploys machine controllers based on the value of the `Infrastructure` resource `status.platformStatus.type`. When installing, the customer will set this to the main topology for the cluster, I expect to match the control plane. In the case of an RHV control plane, we expect the oVirt Machine Controller to be deployed, but not the baremetal machine controller.
> 
That is correct in the mix-env deployment the only machine controller is the oVirt machine controller, meaning that there will be only oVirt machines, but there can be **nodes** that are not oVirt and that is fine.

> So any Machine that is created, is expected to be an oVirt machine.
+1

> A customer might expect to be able to create an oVirt machine and a baremetal machine, but the oVirt controller is the only one running, and it won't understand the baremetal machines so will error on them constantly.
> 
Yes but a user can have a baremetal nodes and the controller will not care.
 
> Machine API isn't designed to work multi platform today, so I don't think there's a way we can deploy both oVirt and baremetal machine controllers simultaneously and expect them to work

Author: I'm switching the focus of this enhancement to be specific to the provider interface and not to add the oVirt provider.
As discussed on other threads as well, the ovirt provider addition is more of a result and maybe will be the first example of this interface implementation

### Goals

> - Create a Provider interface, and hooks in the assisted installer code which separates the assisted installer core code

Reviewer 2: It would be an improvement to rephrase this as an outcome rather than a solution. The goal is to enable additional platforms in some way, so that a user can create a cluster with platform integration. What you're describing here is solution details that would be better in the Proposal section.

Author: I aimed for the [provider interface](https://github.com/openshift/assisted-service/pull/2272/files#diff-988c4e5d655472b2174837171f5d89978ca0f86ec8554329b386dccb9990ea83R84-R90) to do that.
I would expand it and update the enhancement as we will find more provider-specific parts, in the meantime, I relaied on the adding vSphere PR that was merged by the team

Author: OK I agree, moved to the proposal part and replaced with:
- Define a clear way to add a specific platform provider to the assisted installer.
- Add the oVirt platform provider to the assisted installer.
- Migrate all existing supported providers to implement the same unified interface.

Is that better?

 from the provider-specific code.
- Add clear documentation for each provider specific function.
- Document steps to install a provider cluster with assisted installer

### Non-Goals

- Support IPI installations.

> - Integrating all relevant [Openshift installer types](https://github.com/openshift/installer/tree/master/pkg/types)

Reviewer: ```suggestion
- Integrating any relevant [Openshift installer types](https://github.com/openshift/installer/tree/master/pkg/types)
```

into the assisted installer code - this will require a large refactor of the assisted installer code.

> - Remove Baremetal from not appropriate struct/function names (example in open questions).

Reviewer 6: Please add:
 ```suggestion
- Remove Baremetal from not appropriate struct/function names (example in open questions).
- Adding Support for cloud installations (AWS/GCP/Azure)
```

Reviewer: I think that adding support for any platform is a non-goal, and I think that's what @Gal-Zaidman meant in the second bullet "Integrating all relevant Openshift installer types".

Author: No, I actually meant importing types from the installer project, we noticed that there are a lot of types that are duplicated from the installer project in the assisted installer, and I meant we don't intend to remove that duplication but go with the same implementation and add more structs that are essentially mirroring the structs in the installer project.

 
## Proposal

### User Stories

#### Story 1

> As a RHV admin I want to create an Openshift cluster on my RHV engine by using the assisted installer.

Reviewer 2: This is possible today. I think the difference you're seeking is that you want the resulting cluster to have RHV as its platform so it can take advantage of day-2 platform integrations?

#### Story 2

As a RHV admin I want to create an Openshift cluster on my RHV engine by using the assisted installer in which
Control-Plain (Master) nodes are deployed on top of RHV and Compute (Worker) nodes are Bare-Metal nodes.

#### Story 3

As an Openshift developer working on platform X I want to add support for installing Openshift with assisted installer

> on top of platform X.

Reviewer 4: This is something that will be possible when this work is done. However, I wonder if we should make it an explicit part of the stories supporting this enhancement. There's a discussion with regards to only supporting `on-prem` and not cloud providers but, as a developer reading this enhancement, I may feel encouraged to add support for cloud providers too.

Am I missing something?

Reviewer: The whole point of creating this generic infrastructure is to make it easier to support more platforms.  I guess that's the point of this story.

Author: We talked about it above, I don't think it encourages you to implement cloud platforms, but just doesn't talk about it.
The interface which we will define is about adding a provider, there is probably nothing to specific for cloud provider but adding any provider will require the approval and support of the assisted installer team.
We do mention adding cloud providers as a non-goal of the current enhancement if it helps

Implement the registry pattern so providers can register with their names.
When we will detect the a certain provider then we will register it.
If no provider is detected then we register the Bare-Metal provider.

##### Step 4: Implement providerDetector method

Implement a function which determines which providers can be used.
Each provider should implement its own detection logic.
For example:
if all control plain nodes are running on oVirt
(i.e have [SystemVendor.Manufacturer](https://github.com/openshift/assisted-service/blob/master/models/system_vendor.go)
== oVirt OR RHV) then register the oVirt provider.

##### Step 5: Add calls to the provider functions

Whenever we need to call some provider-specific function we should use the provider interface instead of a switch case.

##### Step 6: Implement Default Provider

Will contain default implementation of most functions required by Provider interface. 
It should be embedded in all providers to make adding new APIs easier.

##### Step 7: Implement the oVirt provider and migrate Baremetal and vSphere providers

Implement the oVirt provider and migrate existing providers to the Provider interface.

> #### Changes to the assisted-agent project:

Reviewer 3: We need another interface for validations .. some will be per provider

Author: Yes you are right, I'll add flows added on https://github.com/openshift/assisted-service/pull/2274

Since the openshift installer is generating a random InfraID the names of the Nodes(in the manifest) are different
from the nodes the user provisions, and since we have a provider set then the provider will try to start new machines 
due to the naming mismatch.
Since the InfraID is random we can't tell the user to create the Nodes with specific names.
Also, the InfraID is later used by the platforms in various ways to identify the cluster,
for example in oVirt we use it to set an oVirt tag which helps us group the cluster resources in oVirt,
and we rely on it in various places such as cluster destroy or CSI/Machine provider logic to filter cluster resources
quickly and prevent unnecessary API calls.

We need to add an env var to allow us to override the generated InfraID.
We started a PR on the [installer project](https://github.com/openshift/installer/pull/5101).

### Risks and Mitigations

As with every large code change, this change contains the risk of breaking existing functionality.
This can be mitigated by sanity testing and making sure each provider implements unit tests and e2e test suite.

## Design Details

### Open Questions

> 1. Throught the assisted installer code we have structs and functions that are specific to baremetal(at least from the name),

Reviewer 2: s/Throught/Throughout/

> - Are they safe to be extended to support providers?

Reviewer 6: The platform struct is already extended by the `None` and `Vsphere` platform

Author: yes but there are many of them not only the install config...
I think this is more of a general question to keep in mind as we develop, plus maybe it will make sense to edit some of them to have a general name that is not [SOMETHING]baremetal... not a must but will probably make sense in some places

> - Is there anything baremetal specific we should be conserned about?

Reviewer 2: s/conserned/concerned/

- Is the assisted installer team aims to modify the names to be generic? for example I don't see why the
[InstallerConfig](https://github.com/openshift/assisted-service/blob/master/internal/installcfg/installcfg.go#L75-L110)
 struct is BareMetal.
 
 Some examples but it is on every file:
- [BMACReconciler](https://github.com/openshift/assisted-service/blob/master/internal/controller/controllers/bmh_agent_controller.go#L50-L55)
- [bareMetalInventory](https://github.com/openshift/assisted-service/blob/master/internal/bminventory/inventory.go#L149-L175)
- [InstallerConfigBaremetal](https://github.com/openshift/assisted-service/blob/master/internal/installcfg/installcfg.go#L75-L110)
- [bmhIsMaster](https://github.com/openshift/assisted-service/blob/master/internal/ignition/ignition.go#L541-L551)

> 2. Is there a use case in which the user will provision the Nodes on a certain provider but want to disable

Reviewer 2: That is a use case that seems worth retaining. OpenShift users often do UPI installation because something about the platform integration isn't compatible with their environment or goals. AI aims to make such installation experiences much easier, and sometimes the resulting cluster will deliberately avoid enabling support for a specific platform.

Reviewer 6: Agree, this is why the provider registration should mean that the provider is supported but the user can still choose the platform.

Author: I asked that above in a different conversation, one thing that is not clear to me (due to my lack of AI project knowledge probably) how will the user select the platform if there is no UI changes?

Reviewer: We already have an API that returns the platforms that are suitable to the current hosts, and another API to select a platform and provide its configuration.

provider integration? meaning nodes will be provider will be discovered but we want platform none anyway?
in which case what would be the best way to handle this in terms of user interaction?

> would an env var flag to disable provider integration is enough?

Reviewer 2: This seems like a per-cluster setting, not something we'd enable or disable for the entire application.

Reviewer: Correct.  Currently, with the vSphere integration, we detect the hosts' platforms and if they are all vSphere, we report via API that vSphere platform integration is possible.  However, the user must choose whether to enable it or not.

Author: Probably answers my question above :sweat_smile: 
Is there a vSphere integration demo around?

Reviewer: @eliorerz Can you please provide a link to the demo that you recorded?

- Red Hat Virtualization (oVirt).
- OpenStack.

Currently, the assisted installer supports only the Bare-Metal and vSphere platforms, but for future on-prem
providers we need to make sure there is a clear interface each on prem provider can implement to add its platform.

This enhancement will define a clear interface to extend the assisted installer to add on prem provider integrations.

## Motivation

Assisted Installer can be a great tool to perform User Provisioned installation (UPI)  installation on external providers,
providing the user with clear UI to see its provisioned machines joining the cluster, and lowering installation
resources requirements.
It is only natural that on-prem providers will want to follow vSphere path and extend assisted installer to support
their platform, to make sure that the provider specific code is easy to maintain, doesn't burden the assisted installer team,
and the providers have a clear understanding of what they need to implement to add support to their platform we need to
define a clear interface which each provider can implement and maintain.

### Goals

- Define a clear way to add a specific platform provider to the assisted installer.

> - Migrate all existing supported providers to implement the same unified interface.

Reviewer 10: Test it easily with assisted-test-infra

### Non-Goals

- Support IPI installations.
- Integrating any relevant [Openshift installer types](https://github.com/openshift/installer/tree/master/pkg/types)
into the assisted installer code - this will require a large refactor of the assisted installer code.

> - Remove Baremetal from not appropriate struct/function names.

Reviewer 10: ```suggestion
- Remove "Baremetal" or "BM" words from inappropriate struct/function names. (as this is how the assisted installer was originally implemented)
```

> - Adding Support for cloud installations (AWS/GCP/Azure).

Reviewer 10: I find this bullet similar to the 2nd non-goal, but a bit confusing since specifying only the cloud installations while there are much more.
IIUC, your intention is this enhancement doesn't imply supporting any OpenShift Installer supported providers, regardless if they are cloud providers or not.

Author: done

##### Step 2: Create a Go interface

We should create a Go interface that will contain each provider specific logic in a function
for each provider to implement.

Prototype:

```go
// Provider contains functions which are required to support installing on a specific platform.
type Provider interface {
    // Name returns the name of the platform.
    Name() models.PlatformType
    // AddPlatformToInstallConfig adds the provider platform to the installconfig platform field,
    // sets platform fields from values within the cluster model.
    AddPlatformToInstallConfig(cfg *installcfg.InstallerConfigBaremetal, cluster *common.Cluster) error
    // SetPlatformValuesInDBUpdates updates the `updates` data structure with platform specific values
    SetPlatformValuesInDBUpdates(platformParams *models.Platform, updates map[string]interface{}) error
    // SetPlatformUsages uses the usageApi to update platform specific usages
    SetPlatformUsages(platformParams *models.Platform, usages map[string]models.Usage, usageApi usage.API) error
    // AreHostsSupported checks if the provider supports the hosts from values discovered in the inventory data

>     AreHostsSupported(hostInventories []*models.Inventory) (bool, error)

Reviewer 10: 1. I think a validation for a single host could be easier to manipulate and run in a loop if necessary instead of just supporting arrays.
2. I think the wrapping `models.Host` for the Inventory is better since we can add more business logic metadata to the host without it being detected by the agent and as such sent as part of the Inventory model.

}
```

##### Step 3: Implement registry

Implement the registry pattern so providers can register with their names.
When we will detect that certain provider then we will set it as the current provider in the registry.
If no provider is detected then we register the Bare-Metal provider.

##### Step 4: Implement Default Provider

Will contain default implementation of most functions required by Provider interface.
It should be embedded in all providers to make adding new APIs easier.

> ##### Step 4: Add the provider registry to relevant structures and call it

Reviewer 10: Step number wasn't incremented
